javascript/ql/src/semmle/javascript/Classes.qll

```diff
@@ -153,26 +163,57 @@ class ClassDefinition extends @classdefinition, ClassOrInterface, AST::ValueNode
    */
   override predicate isAbstract() { isAbstractClass(this) }
 
-  override string describe() {
-    if exists(inferNameFromVarDef())
-    then result = inferNameFromVarDef()
-    else result = "anonymous class"
+  override string describe() { inferDescription(result, _) }
+
+  override string getInferredName() {
+    inferDescription(_, result) and
+    result != ""
+  }
+
+  /**
+   * Infers a `description` and `name` of this class.
+   *
+   * If no name could be found, `name` is an empty string.
+   */
+  private predicate inferDescription(string description, string name) {
+    if exists(inferNameFromVarDef(_))
+    then description = inferNameFromVarDef(name)
+    else
+      if exists(inferNameFromProp(_))
+      then description = inferNameFromProp(name)
+      else (
+        description = "anonymous class" and name = ""
+      )
   }
 
   /**
    * Gets a description of this class, based on its declared name or the name
    * of the variable it is assigned to, if any.
    */
-  private string inferNameFromVarDef() {
+  private string inferNameFromVarDef(string name) {
     // in ambiguous cases like `let C = class D {}`, prefer `D` to `C`
     if exists(getName())
-    then result = "class " + getName()
+    then name = getName() and result = "class " + name
     else
       exists(VarDef vd | this = vd.getSource() |
-        result = "class " + vd.getTarget().(VarRef).getName()
+        name = vd.getTarget().(VarRef).getName() and
+        result = "class " + name
       )
   }
 
+  /**
+   * Gets a description of this class, based on the name of a property
+   * it is stored in.
+   */
+  private string inferNameFromProp(string name) {
+    exists(AssignExpr assign, DotExpr lhs |
+      this = assign.getRhs() and
```

Comment:
```suggestion
      this = assign.getRhs().getUnderlyingValue() and
```

Reply:
Good idea. It can lead to ambiguous names in cases like `x.foo = x.bar = function() {}` so I'll withhold resolution until we make a decision on that.